File: 29Items-Projects/7Algorithms/6-Backtracking/python/graph_coloring.py
"""Graph Coloring — assign colors so no adjacent vertices share the same color."""

COLOR_NAMES = ["Red", "Green", "Blue", "Yellow", "Purple", "Orange", "Cyan", "White"]


class GraphColoring:
    def __init__(self, adj: list[list[int]], num_colors: int):
        self.adj = adj
        self.num_colors = num_colors
        self.v = len(adj)
        self.colors = [-1] * self.v

    def solve_and_print(self) -> None:
        print(f"Graph with {self.v} vertices, {self.num_colors} colors")
        print("Adjacency matrix:")
        for row in self.adj:
            print(f"  {row}")
        print()

        success = self._backtrack(0)

        if success:
            print("Valid coloring found:")
            for i, c in enumerate(self.colors):
                name = COLOR_NAMES[c] if c < len(COLOR_NAMES) else f"C{c}"
                print(f"  Vertex {i} → {name}")
        else:
            print(f"No valid coloring exists with {self.num_colors} colors.")

    def _backtrack(self, vertex: int) -> bool:
        if vertex == self.v:
            return True
        for c in range(self.num_colors):
            if self._is_safe(vertex, c):
                self.colors[vertex] = c
                if self._backtrack(vertex + 1):
                    return True
                self.colors[vertex] = -1
        return False

    def _is_safe(self, vertex: int, color: int) -> bool:
        for i in range(self.v):
            if self.adj[vertex][i] == 1 and self.colors[i] == color:
                return False
        return True

File: 29Items-Projects/7Algorithms/6-Backtracking/java/src/graphcoloring/GraphColoring.java
package graphcoloring;

import java.util.*;

public class GraphColoring {

    private final int[][] adj;
    private final int numColors;
    private final int v;
    private final int[] colors;

    public GraphColoring(int[][] adj, int numColors) {
        this.adj = adj;
        this.numColors = numColors;
        this.v = adj.length;
        this.colors = new int[v];
        Arrays.fill(colors, -1);
    }

    public void solveAndPrint() {
        System.out.printf("Graph with %d vertices, %d colors%n", v, numColors);
        System.out.println("Adjacency matrix:");
        for (int[] row : adj) System.out.println(Arrays.toString(row));
        System.out.println();

        boolean success = backtrack(0);

        if (success) {
            System.out.println("Valid coloring found:");
            String[] names = {"Red", "Green", "Blue", "Yellow", "Purple", "Orange", "Cyan", "White"};
            for (int i = 0; i < v; i++) {
                String colorName = colors[i] < names.length ? names[colors[i]] : "C" + colors[i];
                System.out.printf("  Vertex %d → %s%n", i, colorName);
            }
        } else {
            System.out.println("No valid coloring exists with " + numColors + " colors.");
        }
    }

    private boolean backtrack(int vertex) {
        if (vertex == v) return true;

        for (int c = 0; c < numColors; c++) {
            if (isSafe(vertex, c)) {
                colors[vertex] = c;
                if (backtrack(vertex + 1)) return true;
                colors[vertex] = -1;
            }
        }
        return false;
    }

    private boolean isSafe(int vertex, int color) {
        for (int i = 0; i < v; i++) {
            if (adj[vertex][i] == 1 && colors[i] == color) return false;
        }
        return true;
    }
}
